### PR TITLE
Add Kennedy simphydro to ELM

### PR DESCRIPTION
**Pull request by pnlfang**
Adding plant hydraulics model of Kennedy et al. (2019).  

Kennedy, D., S. Swenson, K. W. Oleson, D. M. Lawrence, R. Fisher, A. C. L. da Costa,
and P. Gentine (2019), Implementing Plant Hydraulics in the Community Land Model,
Version 5, J Adv Model Earth Sy, 11(2), 485-513, doi:10.1029/2018MS001500.

[BFB]

### CODE REVIEW

**Comment by thorntonpe** [comment]
@pnlfang  has this PR gone through any testing with the test suites? Is it BFB?

**Comment by rljacob** [comment]
Is this a v2 feature?  Its needs a pointer to a design document.

**Comment by pnlfang** [comment]
> @pnlfang has this PR gone through any testing with the test suites? Is it BFB?

**Comment by pnlfang** [comment]
> 
> 
> > @pnlfang has this PR gone through any testing with the test suites? Is it BFB?

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 1]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+         !crop       => veg_vp%crop                           , & ! Input:  crop or not (0 =not crop and 1 = crop)
+         leafcn     => veg_vp%leafcn                         , & ! Input:  leaf C:N (gC/gN)
+         flnr       => veg_vp%flnr                           , & ! Input:  fraction of leaf N in the Rubisco enzyme (gN Rubisco / gN leaf)
+         fnitr      => veg_vp%fnitr                          , & ! Input:  foliage nitrogen limitation factor (-)
+         slatop     => veg_vp%slatop                         , & ! Input:  specific leaf area at top of canopy, projected area basis [m^2/gC]
+         stem_leaf     => veg_vp%stem_leaf                         , & ! allocation parameter: new stem c per new leaf C (gC/gC)
+         froot_leaf     => veg_vp%froot_leaf                         , & ! allocation parameter: new fine root C per new leaf C (gC/gC)
+         croot_stem     => veg_vp%croot_stem                         , & ! allocation parameter: new coarse root C per new stem C (gC/gC)
+         !i_vcad     => veg_vp%i_vcad                         , & ! Input:  [real(r8) (:)   ]
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 2]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+         leafn_xfer    => veg_ns%leafn_xfer      , &
+         leafp         => veg_ps%leafp         , &
+         leafp_storage => veg_ps%leafp_storage , &
+         leafp_xfer    => veg_ps%leafp_xfer    , &
+         i_vcmax       => veg_vp%i_vc                          , &
+         s_vcmax       => veg_vp%s_vc                          , &
+         bsw           => soilstate_inst%bsw_col                , & ! Input:  [real(r8) (:,:) ]  Clapp and Hornberger "b"
+         sucsat        => soilstate_inst%sucsat_col             ,  & ! Input:  [real(r8) (:,:) ]  minimum soil suction (mm)
+         ivt           => veg_pp%itype                             & ! Input:  [integer  (:)   ]  patch vegetation type
+         !leafresp_method=> photosyns_inst%leafresp_method     , & ! Input:  [integer        ]  method type to use for leaf-maint.-respiration at 25C canopy top
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 3]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+     )
+         !leaf_mr_vcm => canopystate_inst%leaf_mr_vcm           & ! Input:  [real(r8)       ]  scalar constant of leaf respiration with Vcmax         )
+      an_sun        =>    photosyns_inst%an_sun_patch         ! Output: [real(r8) (:,:) ]  net sunlit leaf photosynthesis (umol CO2/m**2/s)
+      an_sha        =>    photosyns_inst%an_sha_patch         ! Output: [real(r8) (:,:) ]  net shaded leaf photosynthesis (umol CO2/m**2/s)
+      par_z_sun     =>    solarabs_inst%parsun_z_patch        ! Input:  [real(r8) (:,:) ]  par absorbed per unit lai for canopy layer (w/m**2)
+      lai_z_sun     =>    canopystate_inst%laisun_z_patch     ! Input:  [real(r8) (:,:) ]  leaf area index for canopy layer, sunlit or shaded
+      vcmaxcint_sun =>    surfalb_inst%vcmaxcintsun_patch     ! Input:  [real(r8) (:)   ]  leaf to canopy scaling coefficient
+      alphapsn_sun  =>    photosyns_inst%alphapsnsun_patch    ! Input:  [real(r8) (:)   ]  13C fractionation factor for PSN ()
+      !o3coefv_sun   =>    ozone_inst%o3coefvsun_patch         ! Input:  [real(r8) (:)   ]  O3 coefficient used in photosynthesis calculation
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 4]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+      psn_wc_sha    =>    photosyns_inst%psnsha_wc_patch      ! Output: [real(r8) (:)   ]  Rubisco-limited foliage photosynthesis (umol co2 /m**2/ s) [always +]
+      psn_wj_sha    =>    photosyns_inst%psnsha_wj_patch      ! Output: [real(r8) (:)   ]  RuBP-limited foliage photosynthesis (umol co2 /m**2/ s) [always +]
+      psn_wp_sha    =>    photosyns_inst%psnsha_wp_patch      ! Output: [real(r8) (:)   ]  product-limited foliage photosynthesis (umol co2 /m**2/ s) [always +]
+      gs_mol_sha    =>    photosyns_inst%gs_mol_sha_patch     ! Output: [real(r8) (:,:) ]  shaded leaf stomatal conductance (umol H2O/m**2/s)
+      vegwp         => canopystate_inst%vegwp_patch           ! Input/Output: [real(r8) (:,:) ]  vegetation water matric potential (mm)
+      !==============================================================================!
+      ! Photosynthesis and stomatal conductance parameters, from:
+      ! Bonan et al (2011) JGR, 116, doi:10.1029/2010JG001593
+      !==============================================================================!
+! calculate root-soil interface conductance
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please indent this comment and multiple comments that are below

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 5]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+            rs_resis = 1._r8/soil_conductance + 1._r8/root_conductance
+
+! conductance is inverse resistance
+! explicitly set conductance to zero for top soil layer
+            if(rai(j)*rootfr(p,j) > 0._r8 .and. j > 1) then
+               k_soil_root(p,j) =  1._r8/rs_resis
+            else
+               k_soil_root(p,j) =  0.
+            endif
+!Fang
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 6]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+                     vcmax25top = min(max(vcmax25top, 10.0_r8), 150.0_r8)
+                     jmax25top = min(max(jmax25top, 10.0_r8), 250.0_r8)
+                  else
+                     lnc(p) = 0.0_r8
+                     lpc(p) = 0.0_r8
+                     vcmax25top = 0.0_r8
+                     jmax25top = 0.0_r8
+                  end if
+
+!                  if (lnc(p) >= 0.1_r8 .and. lnc(p) <=3.0_r8 .and. lpc(p) >= 0.05_r8 .and. lpc(p) <= 0.5_r8) then
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 7]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+                  vcmax_z(p,sha,iv) = vcmax_z(p,sha,iv) / (1._r8 + exp( 0.2_r8*((tfrz+15._r8)-t_veg(p)) ))
+                  vcmax_z(p,sha,iv) = vcmax_z(p,sha,iv) / (1._r8 + exp( 0.3_r8*(t_veg(p)-(tfrz+40._r8)) ))
+               end if
+
+               kp_z(p,sun,iv) = kp25_sun * 2._r8**((t_veg(p)-(tfrz+25._r8))/10._r8)
+               kp_z(p,sha,iv) = kp25_sha * 2._r8**((t_veg(p)-(tfrz+25._r8))/10._r8)
+
+            end if
+
+           ! Change to add in light inhibition of respiration. 0.67 from Lloyd et al. 2010, & Metcalfe et al. 2012
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 8]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+
+         ! Loop through canopy layers (above snow). Only do calculations if daytime
+
+         do iv = 1, nrad(p)
+
+            if (par_z_sun(p,iv) <= 0._r8) then        ! night time
+
+               !zqz temporary signal for night time
+               vegwp(p,1)=1._r8
+!               if (      stomatalcond_mtd == stomatalcond_mtd_bb1987 )then
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 9]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+               !KO  Follow CLM photosynthesis to limit bbb
+               rs_z_sha(p,iv) = min(rsmax0, 1._r8/(max( bsha(p)*bbb(p), 1._r8 )) * cf)
+               ci_z_sha(p,iv) = 0._r8
+               rh_leaf_sha(p) = 0._r8
+
+            else                                     ! day time
+
+               !now the constraint is no longer needed, Jinyun Tang
+               ceair = min( eair(p),  esat_tv(p) )
+!               rh_can = ceair / esat_tv(p)
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 10]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+               end if
+
+               !find ci and stomatal conductance
+               call hybrid_PHS(ci_z_sun(p,iv), ci_z_sha(p,iv), p, iv, c, t, gb_mol(p), bsun(p),bsha(p), je_sun, &
+                               je_sha, cair(p), oair(p), lmr_z_sun(p,iv), lmr_z_sha(p,iv), &
+                               par_z_sun(p,iv), par_z_sha(p,iv), rh_can, gs_mol_sun(p,iv), gs_mol_sha(p,iv), &
+                               qsatl(p), qaf(p), iter1, iter2, atm2lnd_inst, photosyns_inst, &
+                               canopystate_inst, waterstate_inst, soilstate_inst, temperature_inst, waterflux_inst)
+
+!               if ( stomatalcond_mtd == stomatalcond_mtd_medlyn2011 )then
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 11]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+                  gsminsun     = bbb(p)
+                  gsminsha     = bbb(p)
+                  gs_slope_sun = mbb(p)
+                  gs_slope_sha = mbb(p)
+!               end if
+
+               ! End of ci iteration.  Check for an < 0, in which case gs_mol = bbb
+
+               !KO  Follow CLM photosynthesis to limit bbb
+!               if (an_sun(p,iv) < 0._r8) gs_mol_sun(p,iv) = max( bsun(p)*bbb(p), 1._r8 )
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 12]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+
+               if (gs_mol_sun(p,iv) < 0._r8 .or. gs_mol_sha(p,iv) < 0._r8) then
+                  write (iulog,*)'Negative stomatal conductance:'
+                  write (iulog,*)'p,iv,gs_mol_sun,gs_mol_sha= ',p,iv,gs_mol_sun(p,iv),gs_mol_sha(p,iv)
+                  call endrun(decomp_index=p, clmlevel=namep, msg=errmsg(__FILE__, __LINE__))
+               end if
+
+               ! Compare with Ball-Berry model: gs_mol = m * an * hs/cs p + b
+
+               hs = (gb_mol(p)*ceair + gs_m
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines

**Comment by jqyin** [comment]
@pnlfang just to follow up, is this PR bit-for-bit against the test suites?

**Comment by pnlfang** [comment]
@jqyin yes

**Comment by rljacob** [comment]
@bishtgautam please re-review to see if your requested changes were made.

File: components/clm/src/biogeophys/PhotosynthesisMod.F90 [hunk 13]
```diff
@@ -1438,5 +1570,2588 @@ real(r8)             , intent(out)   :: fval     ! return function of the value
...
+         do iv = 1, nrad(p)
+
+            if (par_z_sun(p,iv) <= 0._r8) then        ! night time
+
+               !zqz temporary signal for night time
+               vegwp(p,1)=1._r8
+!               if (      stomatalcond_mtd == stomatalcond_mtd_bb1987 )then
+                  gsminsun = bbb(p)
+                  gsminsha = bbb(p)
+!               else if ( stomatalcond_mtd == stomatalcond_mtd_medlyn2011 )then
```
Comment (bishtgautam) [inline components/clm/src/biogeophys/PhotosynthesisMod.F90]: Please delete the commented out lines.

**Comment by bishtgautam** [comment]
The requested changes from previous review have not been made. I have added a request for one more change in the latest review.

**Comment by rljacob** [comment]
@jqyin please start merging this.

**Comment by rljacob** [comment]
Nevermind, still needs a code review document approval from @kvcalvin

**Comment by rljacob** [comment]
Reassigning to balance load.

**Comment by rljacob** [comment]
@thorntonpe can you please finish the review for this?

**Comment by rljacob** [comment]
@bishtgautam looks like this can finally be merged.

**Comment by bishtgautam** [comment]
Today I re-merged #3107, so I will merge this PR tomorrow.

**Pull request merged**